### PR TITLE
[istio] Removing wildcard RBACs from the module

### LINKED ISSUE #8517: Minimize RBAC permissions of platform components
### Preflight Checklist

- [X] I agree to follow the [Code of Conduct](https://github.com/deckhouse/deckhouse/blob/main/CODE_OF_CONDUCT.md) that this project adheres to.
- [X] I have searched the [issue tracker](https://github.com/deckhouse/deckhouse/issues) for an issue that matches the one I want to file, without success.

### Use case. Why is this important?

Kubernetes CIS [we recommend not using wildcard rights in Role/ClusterRole](https://github.com/aquasecurity/kube-bench/blob/0f8dfaf1156652b556c913aa8e78fccbb1dfd186/cfg/rke-cis-1.24/policies.yaml#L29-L35). 
However, the DKP platform does not take this recommendation into account everywhere. 
For example:
- dex has [verbs: *](https://github.com/deckhouse/deckhouse/blob/main/modules/150-user-authn/templates/dex/rbac-for-us.yaml#L53)
- upmeter-agent [has wildcard rights to all resources](https://github.com/deckhouse/deckhouse/blob/main/modules/500-upmeter/templates/upmeter-agent/rbac-for-us.yaml#L63-L73) in d8-upmeter namespace

Therefore, the DKP does not fully comply with the Kubernetes CIS recommendations.

### Proposed Solution

Grant specific required permissions to objects in the Kubernetes API.

### Additional Information

You might be interested in the [audit2rbac](https://github.com/liggitt/audit2rbac) tool to minimize permissions for the modules.

### PR DESCRIPTION
**Pull request by Beastlex**
## Description
Removing wildcards from ClusterRoles for istio operator and control-plane. Discharging istio-operator from deploying RBAC.

Fixes #8517

## Why do we need it, and what problem does it solve?
Setting a wildcard for Role/ClusterRole is potentially dangerous. We will create RBACs using the platform instead of using the istio-operator, which deploys these roles with wildcards.

## Checklist
- [x] The code is covered by unit tests.
- [ ] e2e tests passed.
- [ ] Documentation updated according to the changes.
- [x] Changes were tested in the Kubernetes cluster manually.

## Changelog entries

```changes
section: istio
type: chore
summary: Removed wildcard RBAC from istio-operator and control-plane. istio-operator discharged from deploying RBACs, we handle them from our templates.
impact_level: default
```

### CODE REVIEW

File: modules/110-istio/templates/control-plane/_iop_clusterroles.tpl [hunk 1]
```diff
@@ -0,0 +1,360 @@
+{{- define "iop_main_clusterrole_rules" -}}
```
Comment (apolovov) [inline modules/110-istio/templates/control-plane/_iop_clusterroles.tpl:1]: _istiod_clusterroles.tpl

File: modules/110-istio/templates/control-plane/iop.yaml [hunk 2]
```diff
@@ -285,4 +285,35 @@ spec:
       - matchExpressions:
         - key: job-name
           operator: Exists
+---
```
Comment (apolovov) [inline modules/110-istio/templates/control-plane/iop.yaml:288]: create a folder iop/istiod and there will be the rbac-for-us.yaml  template.

File: modules/040-control-plane-manager/hooks/audit_policy.go [hunk 3]
```diff
@@ -52,6 +53,17 @@ var _ = sdk.RegisterFunc(&go_hook.HookConfig{
 			},
 			FilterFunc: filterAuditSecret,
 		},
+		{
+			Name:       "extra_configmaps_with_serviceaccaunts",
+			ApiVersion: "v1",
+			Kind:       "ConfigMap",
+			LabelSelector: &apiv1.LabelSelector{
+				MatchLabels: map[string]string{
+					"deckhouse.io/extra-configs": "serviceaccounts",
```
Comment (apolovov) [inline modules/040-control-plane-manager/hooks/audit_policy.go:62]:
```suggestion
					"control-plane-manager.deckhouse.io/extra-audit-policy-config": "",
```

File: modules/040-control-plane-manager/hooks/audit_policy.go [hunk 4]
```diff
@@ -68,11 +80,29 @@ func filterAuditSecret(unstructured *unstructured.Unstructured) (go_hook.FilterR
...
+func filterConfigMap(unstructured *unstructured.Unstructured) (go_hook.FilterResult, error) {
+	var cm v1.ConfigMap
+	serviceAccounts := make([]string, 0)
+
+	err := sdk.FromUnstructured(unstructured, &cm)
+	if err != nil {
+		return nil, err
+	}
+
+	if data, ok := cm.Data["serviceAccounts"]; ok {
```
Comment (apolovov) [inline modules/040-control-plane-manager/hooks/audit_policy.go:92]:
```suggestion
	if data, ok := cm.Data["basicAuditPolicy"]; ok {
```

```
data:
  basicAuditPolicy: |
    serviceAccounts:
    - system:serviceaccount:d8-ingress-nginx:ingress-nginx
    - system:serviceaccount:d8-cloud-provider-yandex:cloud-metrics-exporter
```

File: modules/040-control-plane-manager/hooks/audit_policy.go [hunk 5]
```diff
@@ -52,6 +53,17 @@ var _ = sdk.RegisterFunc(&go_hook.HookConfig{
 			},
 			FilterFunc: filterAuditSecret,
 		},
+		{
+			Name:       "extra_configmaps_with_serviceaccaunts",
```
Comment (apolovov) [inline modules/040-control-plane-manager/hooks/audit_policy.go:57]:
```suggestion
			Name:       "configmaps_with_extra_basic_policy",
```

File: modules/110-istio/hooks/migrate_serviceaccounts.go [hunk 6]
```diff
@@ -0,0 +1,104 @@
+/*
```
Comment (apolovov) [inline modules/110-istio/hooks/migrate_serviceaccounts.go:1]: migration_serviceaccounts.go

Add a comment "TODO: remove after 1.64"

File: modules/110-istio/hooks/migrate_serviceaccounts.go [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+)
+
+const (
+	labelHelmManagedBy         = "app.kubernetes.io/managed-by"
+	annotationReleaseName      = "meta.helm.sh/release-name"
+	annotationReleaseNamespace = "meta.helm.sh/release-namespace"
+)
+
+var _ = sdk.RegisterFunc(&go_hook.HookConfig{
+	Queue:        lib.Queue("migrate-service-accounts"),
```
Comment (apolovov) [inline modules/110-istio/hooks/migrate_serviceaccounts.go:39]:
```suggestion
	Queue:        lib.Queue("migration-adopt-service-accounts"),
```

File: modules/110-istio/hooks/migrate_serviceaccounts_test.go [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+package hooks
+
+import (
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+
+	. "github.com/deckhouse/deckhouse/testing/hooks"
+)
+
+var _ = Describe("Istio hooks :: migrate_service_accounts ::", func() {
```
Comment (apolovov) [inline modules/110-istio/hooks/migrate_serviceaccounts_test.go:26]: migration_adopt_service_accounts

File: modules/110-istio/templates/control-plane/iop/istiod/_istiod_clusterroles.tpl [hunk 9]
```diff
@@ -0,0 +1,10 @@
+{{- define "istiod_main_clusterrole_rules" -}}
```
Comment (apolovov) [inline modules/110-istio/templates/control-plane/iop/istiod/_istiod_clusterroles.tpl:1]:
```suggestion
{{- define "istiod_rbac" -}}
```

File: modules/110-istio/templates/operator/rbac-for-us.yaml [hunk 10]
```diff
@@ -13,51 +13,97 @@ metadata:
...
-      - '*'
-  - apiGroups:
-      - config.istio.io
-    resources:
-      - '*'
-    verbs:
-      - '*'
   - apiGroups:
       - install.istio.io
     resources:
```
Comment (apolovov) [inline modules/110-istio/templates/operator/rbac-for-us.yaml:18]: also discharge from updating clusterroles/roles.

File: testing/matrix/linter/rules/roles/placement.go [hunk 11]
```diff
@@ -157,6 +157,11 @@ func objectRBACPlacementServiceAccount(m utils.Module, object storage.StoreObjec
 			}
 			return errors.EmptyRuleError
 		}
+
+		if strings.HasPrefix(objectName, "istiod") && namespace == "d8-istio" {
+			return errors.EmptyRuleError // SAs with predefined names for istiod are generated by operator
```
Comment (apolovov) [inline testing/matrix/linter/rules/roles/placement.go:162]: don't understand.

File: testing/matrix/linter/rules/roles/placement.go [hunk 12]
```diff
@@ -157,6 +157,11 @@ func objectRBACPlacementServiceAccount(m utils.Module, object storage.StoreObjec
 			}
 			return errors.EmptyRuleError
 		}
+
+		if strings.HasPrefix(objectName, "istiod") && namespace == "d8-istio" {
+			return errors.EmptyRuleError // ServiceAccounts are generated by the istio-operator according to its naming conventions
```
Comment (apolovov) [inline testing/matrix/linter/rules/roles/placement.go:162]:
```suggestion
                        // istiod Deployment is rendered by istio-operator with serviceAccountName according to its
                        // naming conventions we can't change (i.e. istiod-v1x19).
                        // In our convention it has to be named as "iop" according to template folder, but within the folder we render
                        // not a single istiod instance, but several for different versions and can't use the shared ServiceAccount for them.
			return errors.EmptyRuleError
```

**Review by apolovov — APPROVED** [review]

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @Beastlex. I've started the workflow for label 'e2e/run/yandex-cloud'...

**Review by name212 — APPROVED** [review]

**Pull request merged**